### ISSUE TITLE
Draw function is not reset when Restart is pressed

### ISSUE THREAD

**Issue by petercollingridge** [Khan/live-editor#588]
If you have a program that starts with no `draw` function defined, but later define one, then when you click Restart, the program will start with the draw function defined. I've made a simple demonstration:
https://www.khanacademy.org/computer-programming/spin-off-of-click-a-button-any-button/6142122579

The program starts without a `draw` function, just drawing some text. When you click it defines a `draw` function. If you press Restart, the program starts with the newly defined `draw` functions. It's a weird way to arrange the code, but it was confusing the creator of the code that I spun-off from into thinking the Restart button was broken.

**Comment by MatthiasSaihttam** [comment]
I think I've figured out how to fix the code, but I'm not sure this is the best way of doing it.
Inserting `this.processing.draw = this.DUMMY;` [here](https://github.com/Khan/live-editor/blob/95b4626f11eaedbadbc21f87b0ea684edcd63aae/js/output/pjs/pjs-code-injector.js#L302) resets the draw function back to the dummy it is initialized to. While this works just fine so far as I can tell, it makes me wonder if we should be resetting all of the PJS callbacks, or even user variables.

**Comment by kevinbarabash** [comment]
It's an interesting bug.  There's an implicit assumption in the code that `draw` will global.  We have a function called `hasOrHadDrawLoop`, but if `draw` is created inside a function this function always returns `false`.  

@MatthiasSaihttam the fix you proposed may indeed be the fix.  Can you try running the tests?

**Comment by petercollingridge** [comment]
Someone made a spin-off of my program showing how you can make variables that are reset on Restarting:
https://www.khanacademy.org/computer-programming/proof-of-concept-restart-proof-variables/6166750996

**Comment by kevinbarabash** [comment]
data shouldn't be persisted across restarts so that program will stop working if/when we fix this issue.

**Comment by EricBalingit** [comment]
Why does not Restart scrap the current program, re-parse the code and rebuild the output from scratch?  Granted with a larger body of code this can take a few seconds, but I would assume that anything live is moot when the restart button is pressed.  Seems like reset should be just that, send a message, die, get reborn.

**Comment by MatthiasSaihttam** [comment]
When I get around to it, I'm going to make the change so that all the PJS callbacks are reset, and I'll have to look into how the environment itself is created and if it's worth it to re-create it from scratch. Right now all restart does is re-run the code, which normally resets variables to what they were initialized to.

**Comment by kevinbarabash** [comment]
Yeah, we could definitely do more when doing a reset.

**Comment by DozenalFTW** [comment]
A few problems though:
1. Some people (like me) like saving variables in `Program` so that they are preserved if you Restart.
1. People should not be doing things like that with the `draw` loop. It is much better to use if statements and `noLoop()`s.

**Comment by kevinbarabash** [comment]
@DozenalFTW I can definitely see how storing variable across restarts is useful.  As for "not doing things like that with the `draw` loop"... live-editor does a lot of things that aren't best practices in order to implement a live updating of code.

**Comment by DozenalFTW** [comment]
No, I meant _users_ should not put `draw` inside other functions.
